### PR TITLE
Centralize White Label Data

### PR DESCRIPTION
**Pull request by davidagustin**

### CODE REVIEW

**Review by lexholden — APPROVED** [review]
LGTM. This is much cleaner!

**Comment by davidagustin** [comment]
I can sure use a manual verify also to see if I covered all the areas

File: app/utils/whitelabel.ts
Comment (lexholden) [inline app/utils/whitelabel.ts:13]:
```suggestion
const configurations: any = {};
```

Hmm good point actually just tried this locally, and it errors because this object is undefined, causing a `Uncaught TypeError: Cannot set property 'title' of undefined` error

Comment (lexholden) [reply app/utils/whitelabel.ts:13]: When I fix that locally, everything looks OK to me. Can probably do some more thorough QA on staging after this is merged

**Review by lexholden — CHANGES_REQUESTED** [review]

File: app/utils/whitelabel.ts
Comment (richardxia) [reply app/utils/whitelabel.ts:13]: I would highly recommend that we don't give this a type of `any`, since that disables type checking on this object, which in turn disables type checking of the entire `whiteLabel` object being exported to other files.

I would suggest that we "transpose" the structure of this object so that the top-level properties are just the different site names (e.g. `SFFAMILIES_DOMAIN`, or in my other suggestion above, `SFFamilies`) and that all the other properties are nested under the site. This would allow us to define a type or interface that describes all of the properties that need to be set for each site, and it would allow the compiler to yell if we're missing any required properties.

The issue with the current structure is that if we added another site, it would be pretty easy to forget to add the new site key to each of the individual top-level properties.

An example of what I'm suggesting is the following:

```ts
interface WhiteLabelSite {
  title: string;
  siteUrl: string;
  appImages: {
    // Fill out each image here
  };
}

const configurations: Record<WhiteLabelSiteKey, WhiteLabelSite> = {};

configurations.SFFamilies = {
  title: 'SF Families',
  siteUrl: 'https://sffamilies.sfserviceguide.org/',
  appImages: { /* fill out images here */ }
}
```

This gives you type safety, preventing you from forgetting or typo'ing any properties, and it'll have better IDE hover information.

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:37]: This icon stuff doesn't look like it has anything to do with whitelabeling, since these are generic to all whitelabeled sites. Could you move this back to `assets/index.js`?

As an aside, I think we should probably get rid of this stuff at some point, since it's not really necessary in a modern frontend project, since Webpack should be able to take care of all the assets transformation stuff, so we really only need to just import the individual icons that we're using like any other Webpack image import.

**Review by richardxia — COMMENTED** [review]
I think this is a good step, but I'd highly recommend that we don't basically turn off all of the TypeScript type checking on `whitelabel.ts`, since the combination of `@ts-ignore` and the `: any` type annotations basically results in almost every variable in that file being untyped.

I'd also suggest we fully eliminate the `isSFFamiliesSite` variable by replacing it with the actual feature that we are gating for each conditional it's used in.

File: app/utils/whitelabel.ts
Comment (davidagustin) [reply app/utils/whitelabel.ts:11]: Great reviews! I become a better engineer everytime. Will push up changes

File: app/pages/HomePage/components/ResourceCard/ResourceCard.jsx
Comment (davidagustin) [inline app/pages/HomePage/components/ResourceCard/ResourceCard.jsx:52]: Another Icon was declared before

File: app/utils/whitelabel.ts
Comment (davidagustin) [inline app/utils/whitelabel.ts:30]: I want to make sure that this is the central place to modify whitelabel data so thus I made this read only.

File: app/utils/whitelabel.ts
Comment (davidagustin) [inline app/utils/whitelabel.ts:85]: Also I want to make sure none of the objects are modified at run time to also encourage development in this file

File: app/pages/ServiceListingPage.tsx
Comment (davidagustin) [inline app/pages/ServiceListingPage.tsx:47]: Collision of variable name here

File: app/components/ui/Navigation/Navigation.jsx
Comment (jjfreund) [inline app/components/ui/Navigation/Navigation.jsx:90]: showMobileNav or something similar would be the configurable property in this case -- true for everyone but sffamilies.

**Review by jjfreund — CHANGES_REQUESTED** [review]

File: app/components/search/SearchEntry.jsx
Comment (davidagustin) [reply app/components/search/SearchEntry.jsx:52]: ok

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [reply app/components/ui/Navigation/Navigation.jsx:56]: We can refactor this on line 61 actually. Will see what I can do.

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [reply app/components/ui/Navigation/Navigation.jsx:90]: We can add a new label, will wait for other reviews too.

File: app/App.tsx
Comment (davidagustin) [reply app/App.tsx:160]: I see, I'd have to implement the white label interface with added properties specifically for SF Families.

File: app/App.tsx
Comment (davidagustin) [reply app/App.tsx:170]: I'm trying to keep the whitelabel interface as consistent as possible while keeping them unaware of each other. I'll see what solutions I can find.

File: app/App.tsx
Comment (davidagustin) [reply app/App.tsx:173]: Ok

File: app/App.tsx
Comment (richardxia) [reply app/App.tsx:160]: UserWay is not a chat widget; it's an accessibility widget. We may potentially want to have both widgets enabled simultaneously, so I think these really should be completely orthogonal properties.

File: app/components/Texting/Texting.jsx
Comment (richardxia) [inline app/components/Texting/Texting.jsx:12]: Why are you importing the `icon` function from the whitelabel instead of `assets`? None of the icons appear to be different for the different whitelabeled sites, so it doesn't need to be part of the whitelabel site. This comment applies to many of the other files in this PR diff, since a lot of them only import the `icon` function and nothing from the actual whitelabel configuration.

File: app/App.tsx
Comment (richardxia) [inline app/App.tsx:11]: Could you name default imports the same as the file being imported, e.g. name this `whitelabel`, not `configurations? That will reduce cognitive load by not introducing a new name. If you don't think the name `whitelabel` is good, then we may want to rename that file.

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:19]: I'm actually a little bit confused by this comment, since all of the whitelabel configurations use the same image for both `logoLarge` and `logoSmall`. What did you mean by "usually different"?

File: app/utils/whitelabel.ts
Comment (richardxia) [reply app/utils/whitelabel.ts:30]: You don't need to mark `WhiteLabelSiteKey` as `Readonly`, since they're strings, and strings are immutable in JavaScript.

It also might be cleaner to put this at the bottom of the file so that you don't have the `Partial<>` anymore, and so that you can just mark the whole thing as readonly. For example, if you did:

```ts
const SFFamilies = {
  title: 'SF Families',
  // etc
} as const;
const SFServiceGuide = { /* stuff goes here */ } as const;
const default = { /* stuff goes here */ } as const;

const configurations: Readonly<Record<WhiteLabelSiteKey, WhiteLabelSite>> = {
  SFFamilies,
  SFServiceGuide,
  default,
} as const;
```

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:35]: Nit: Could you swap the order of the operands across the `===` operator, i.e. put the `window.location.host` on the left-hand side of the `===`? Even though it's logically equivalent, it's much more common in programming and in English to put the thing that is more likely to change on the left-hand side.

File: cypress/support/index.ts
Comment (richardxia) [inline cypress/support/index.ts:22]: What is this for? If this is silencing uncaught exceptions, that this seems like it's a bad thing to do, since we don't want these to slip past CI.

**Review by richardxia — COMMENTED** [review]
Thanks for the updates! I left some additional comments, but I agree with everything that @jjfreund wrote about eliminating all logic checking for the current site outside of the whitelabel.ts file and replacing them with distinct properties.

File: cypress/support/index.ts
Comment (davidagustin) [reply cypress/support/index.ts:22]: ok, yeah this was a workaround for a previous error already solved

File: app/utils/whitelabel.ts
Comment (davidagustin) [reply app/utils/whitelabel.ts:35]: ok

Comment (davidagustin) [reply app/utils/whitelabel.ts:35]: ok for the WhiteLabelSiteKey
default is a reserved word so Id have to change the name

Comment (richardxia) [reply app/utils/whitelabel.ts:35]: Maybe just call it `"AskDarcel"`, since that's what it actually is?

File: app/utils/whitelabel.ts
Comment (davidagustin) [inline app/utils/whitelabel.ts:43]: using "as const" at the end of these objects gives an error in which there needs to be a semi colon before "as const" for example the linter would request something like: {}; as const

File: .eslintrc.js
Comment (davidagustin) [inline .eslintrc.js:28]: These were needed to get passed linter errors for the whitelabel.ts files

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [inline app/components/ui/Navigation/Navigation.jsx:64]: I was being creative here by using regex, hopefully this makes sense but it is a string matching the first portion that must be http:// or https:// to return true.

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [inline app/components/ui/Navigation/Navigation.jsx:97]: mobile nav whiteLabel as requested here

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [inline app/components/ui/Navigation/Navigation.jsx:61]: White label request fulfilled here

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [inline app/components/ui/Navigation/Navigation.jsx:66]: White label for the css of the navLogo here

File: app/utils/whitelabel.ts
Comment (davidagustin) [inline app/utils/whitelabel.ts:83]: default is a reserved word so I had to change it do defaultWhiteLabel

File: app/utils/whitelabel.ts
Comment (davidagustin) [inline app/utils/whitelabel.ts:31]: @fredciaramaglia Hopefully the variable names makes sense. Let me know if there are any questions.

File: app/App.tsx
Comment (davidagustin) [inline app/App.tsx:164]: Orthogonalized userway and intercom, hopefully this is the way that is wanted.

@fredciaramaglia @jjfreund @richardxia

File: app/assets/index.js
Comment (davidagustin) [inline app/assets/index.js:8]: Icon is the only asset imported now because the appImages are moved to the whiteLabel file

**Review by davidagustin — COMMENTED** [review]
Requesting QA on staging @fredciaramaglia @jjfreund @richardxia . Hopefully I got the whitelabel options correct. Will need a thorough check.

File: app/utils/whitelabel.ts
Comment (davidagustin) [reply app/utils/whitelabel.ts:31]: More in the actual file itself, didn't highlight them all.

File: app/utils/whitelabel.ts
Comment (davidagustin) [reply app/utils/whitelabel.ts:43]: @richardxia

**Review by jjfreund — APPROVED** [review]
lgtm -- will merge and test. Further cleanup in subsequent PRs if need be.

**Comment by lgarofalo** [comment]
Looks like a good refactor, I am going to merge and do verification/testing in staging, we won't push a prod release until we have gone through this well.  We will also be adding the linksf config on top of here before prod.

File: app/App.tsx
Comment (richardxia) [reply app/App.tsx:164]: Yep, the way they are now orthogonal looks great!

But I'm also now seeing something a bit weird about these lines, which is that `SFFAMILIES` has still leaked into the line via the `config. SFFAMILIES_USERWAY_APP_ID` configuration setting. This isn't good, because it means that configuration option is actually still tied to the whitelabeled site. At the configuration option level, I think we need to keep the `SFFAMILIES_` prefix because the ID is tied to them and possibly to some subscription they've paid for. However, I'm now thinking that we could maybe handle this within the whitelabel file itself by combining it with the existing `userWay` property.

What do you think of the following idea: In whitelabel.ts, replace the `userWay` property with `userWayAppID`, and have it be `null` either when the whitelabeled site should not have UserWay or when the configuration option is missing? This allows it to play both the role of only enabling it based on which whitelabel site as well as providing the actual value. I think we could do the same thing with Intercom, in which case you could then eliminate the extra check on `config.INTERCOM_APP_ID` here.

File: app/components/ui/Navigation/Navigation.jsx
Comment (richardxia) [reply app/components/ui/Navigation/Navigation.jsx:64]: I think this is a reasonable thing to do, but it might be more readable if you extracted it out into its own function and gave it a name that was more descriptive, like `isExternalURL`.

File: app/components/ui/Navigation/Navigation.jsx
Comment (richardxia) [inline app/components/ui/Navigation/Navigation.jsx:71]: For consistency, you should also set the `to` prop here to `logoLinkDestination`, which would better mirror what's going on in the other case.

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:45]: This comment would be better to place on the interface definition, right above the actual `logoLinkDestination`, since that interface should be the one place where we define what the interface is.

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:110]: I'm not sure if this comment provides enough value to justify its existence. It mostly just describes what `Readonly` means in TypeScript without anything that is novel or specific to this codebase. I would argue that the primary purpose of `Readonly` is to do exactly what you've written here, so repeating that here in the form of a comment doesn't add much information, and possibly worse, it draws extra attention for something that is ordinary.

Do you think you could remove this comment?

File: app/utils/whitelabel.ts
Comment (richardxia) [reply app/utils/whitelabel.ts:117]: Similar to my other PR comment, I don't think this code comment is really saying anything that isn't already obvious from the code just below it. Could you remove it?

File: app/utils/whitelabel.ts
Comment (richardxia) [inline app/utils/whitelabel.ts:119]: You don't need the `?` anymore, since this should never be null or undefined.

Also, even previously, when `whiteLabel` was a `Partial<Record>`, you probably did not want this to use `?`, since that operator should only be used in the case where you want to short-circuit and return null/undefined early. Here, it should be illegal to encounter a null or undefined value, so you should not write code that handles that case, and instead it would be better assert that the value is not null.

Note that this comment doesn't apply to the current version of the code, since `whiteLabel` is no longer a `Partial<Record>`, but in the future if you have code where it should be impossible to have a null value but TypeScript doesn't know about it, you should assert the correct type, not work around a type that should not be possible.

**Review by richardxia — COMMENTED** [review]
Sorry that I'm late to the review, but this LGTM and I don't see any blockers here.

I did leave some comments on a few things that I think would be good to fix sooner rather than later, so if you have time, I think it'd be great if you could fix them in a followup PR, @davidagustin. If you're busy, let me know, and I can submit a PR to fix it up as well.

File: app/components/ui/Navigation/Navigation.jsx
Comment (davidagustin) [reply app/components/ui/Navigation/Navigation.jsx:71]: Acknowledged. Good Ideas, I agree with all of them. Implementing.

**Pull request merged**
